### ISSUE TITLE
`builtins.fetchGit` causes Nix to appear to hang

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#5863]
**Describe the bug**

`git` can take a long time to fetch, particularly on first clone.
This makes `git`'s diagnostic output relevant as a progress indicator.

An example of this problem: https://stackoverflow.com/questions/70561914/nix-shell-script-does-nothing-when-using-script?noredirect=1#comment124765140_70561914

**Steps To Reproduce**

1. Clone fresh repo with `fetchGit` (or clear the cached repos)
2. Observe no output for a significant amount of time

**Expected behavior**

Nix forwards the stderr of git to the user.

If verbosity is a concern, it could be buffered for a couple of seconds and discarded when the process completes successfully before then.

**`nix-env --version` output**

2.5.1 (from the SO question)

**Comment by thufschmitt** [comment]
Agreed. I’m not utterly fond of showing the git output (wouldn’t be very pretty), but maybe we can get git to output in a machine-readable format and use that in the progress bar

**Comment by roberth** [comment]
A mere spinner that progresses for each unbuffered read from the log would already be an improvement.

For a more informative solution, you could parse the one or two ANSI cursor codes that it emits (if not just \r) and turn that into lines that you can feed into the progress bar's build log code.

**Comment by bjornfor** [comment]
Not stale. (Just commenting to prevent auto-close.)

**Comment by armeenm** [comment]
Not stale.

**Comment by minimapletinytools** [comment]
Definitely running into this issue. Adding -v shows `fetching Git repository 'https://github.com/NixOS/nixpkgs/'...` however there is no progress indicator. Wireshark shows it's indeed downloading from git. I ended up cancelling after several minutes and switch to using a tarball instead.

**Comment by jabdoa2** [comment]
Is there any workaround or tutorial on how to pin without waiting forever? https://nixos.wiki/wiki/FAQ/Pinning_Nixpkgs even recommends this (more or less) broken way. For me it is also unclear how to get the sha256 for fetchTarball and how to keep it updated. Is there any tutorial I am missing?

**Comment by roberth** [comment]
@jabdoa2 It's probably best to use a locking tool, such as niv or flakes.
